kratos/factories/standard_linear_solver_factory.h

```diff
@@ -77,13 +78,23 @@ class StandardLinearSolverFactory
      */
     typename LinearSolverType::Pointer CreateSolver(Kratos::Parameters settings) const override
     {
+        auto pinner_solver = typename LinearSolverType::Pointer(new TLinearSolverType(settings));
         if(settings.Has("scaling") && settings["scaling"].GetBool()) {
-            auto pinner_solver = typename LinearSolverType::Pointer(new TLinearSolverType(settings));
+            auto pscaling_solver = typename LinearSolverType::Pointer(new ScalingSolver<TSparseSpace,TLocalSpace>(pinner_solver, true));
+            if (settings.Has("monotonicity_preserving") && settings["monotonicity_preserving"].GetBool()) {
+                return typename LinearSolverType::Pointer(new MonotonicityPreservingSolver<TSparseSpace,TLocalSpace>(pscaling_solver));
+            } else {
+                return pscaling_solver;
+            }
+
+        } else {
+            if (settings.Has("monotonicity_preserving") && settings["monotonicity_preserving"].GetBool()) {
+                return typename LinearSolverType::Pointer(new MonotonicityPreservingSolver<TSparseSpace,TLocalSpace>(pinner_solver));
+            } else {
+                return pinner_solver;
+            }
+        }
```

Comment:
this is quite the mess
imagine someone wants to add yet another solver here? then this will completely blow up with different options

why don't you do this in the CTor of your solver? (same for the scaling solver btw but different PR)
=> this might also solve your complex issues

Reply:
I understand this is quite dirty (this is a draft). But Im not sure what you propose. The point of the scaling_solver is that it works as a wrapper that can contain any other linear solver, I am not sure how you would achieve this in a different way.

Reply:
In a nutshell, the scaling solver would have a call to the LinSolverFactory itself in its ctor, which contains the actual solver

same for this solver here

Reply:
yes, but how do you know you have to call the ScalingSolver ctor? In principle, your solver type is e.g. "amgcl",  not "scaling_solver".

Reply:
ah
yikes

well a proper solution would be to change the input, i.e. the solver type would be scaling, which then has some inner settings (e.g. for amgcl)
then it can be combined arbitrarily

otherwise this will grow completely out of hand the more of those "wrapping" solvers we add I am afraid

Reply:
I agree 100% with philipp on this one.

I think we should construct this as something like 

~~~json
{
      "solver_type" : "MonothonicityPreservingSolver",
      "auxiliary_solver_type" :  {
                        "solver_type" : "AMGCL",
                        "krylov_type": "lgmres",
                        "smoother_type":"damped_jacobi",
                        "preconditioner_type": "relaxation",
                        "verbosity" : 1
       },
      ...here other settings as needed ...
 }
~~~

This should also solve the current compilation problems

Reply:
unrelated but useful:
if you use syntax highlighting for `js` then you can use comments 
@RiccardoRossi I updated your comment

Reply:
thats fine for me (though from interface side, its less clear in my opinion), what would you do with scaling solver? Following the same logic it would be:
```
{
      "solver_type" : "MonothonicityPreservingSolver",
      "auxiliary_solver_type" :  {
             "solver_type" : "ScalingSolver",
             "auxiliary_solver_type" : {
                        "solver_type" : "AMGCL",
                        "krylov_type": "lgmres",
                        "smoother_type":"damped_jacobi",
                        "preconditioner_type": "relaxation",
                        "verbosity" : 1
             }
       }
 }
```
that I think looks much worse than:

```
{
"solver_type" : "AMGCL",
 "krylov_type": "lgmres",
 "smoother_type":"damped_jacobi",
 "preconditioner_type": "relaxation",
 "verbosity" : 1,
 "scaling" : true,
 "monotonicty_preserving" : true
}
```

Reply:
I like the first one more, then I know exactly what is happening and when (I don't know the technical details of those solvers)
this way it would also be straight forward adding more wrapping solvers

Reply:
@ddiezrod you are right, also the scaling solver should be used the same way. 
Actually it is the way it really works. If it is not so it is for "historical reasons".

in any case the nice thing of my (and Philipp's) proposal is that this way it is much more modular